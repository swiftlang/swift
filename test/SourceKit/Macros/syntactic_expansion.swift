//--- test.swift
@DelegatedConformance
@wrapAllProperties
struct Generic<Element> {

  @myPropertyWrapper
  @otherAttr
  var value: Int

  func member() {}
  var otherVal: Int = 1

  #bitwidthNumberedStructs("blah")
}

//--- DelegatedConformance.json
{
  key.macro_roles: [source.lang.swift.macro_role.conformance, source.lang.swift.macro_role.member],
  key.modulename: "MacroDefinition",
  key.typename: "DelegatedConformanceMacro",
}

//--- myPropertyWrapper.json
{
  key.macro_roles: [source.lang.swift.macro_role.accessor, source.lang.swift.macro_role.peer],
  key.modulename: "MacroDefinition",
  key.typename: "PropertyWrapperMacro",
}

//--- wrapAllProperties.json
{
  key.macro_roles: [source.lang.swift.macro_role.member_attribute],
  key.modulename: "MacroDefinition",
  key.typename: "WrapAllProperties",
}

//--- bitwidthNumberedStructs.json
{
  key.macro_roles: [source.lang.swift.macro_role.declaration],
  key.modulename: "MacroDefinition",
  key.typename: "DefineBitwidthNumberedStructsMacro",
}

//--- dummy.script
// REQUIRES: swift_swift_parser

// RUN: %empty-directory(%t)
// RUN: mkdir -p %t/plugins
// RUN: split-file %s %t

//##-- Prepare the macro plugin.
// RUN: %host-build-swift -swift-version 5 -emit-library -o %t/plugins/%target-library-name(MacroDefinition) -module-name=MacroDefinition %S/../../Macros/Inputs/syntax_macro_definitions.swift -g -no-toolchain-stdlib-rpath

// RUN: %sourcekitd-test \
// RUN:   -shell -- echo '### 1' \
// RUN:   == \
// RUN:   -req=syntactic-expandmacro \
// RUN:   -req-opts=1:1:%t/DelegatedConformance.json \
// RUN:   -req-opts=5:3:%t/myPropertyWrapper.json \
// RUN:   -req-opts=2:1:%t/wrapAllProperties.json \
// RUN:   -req-opts=12:3:%t/bitwidthNumberedStructs.json \
// RUN:   %t/test.swift \
// RUN:   -- \
// RUN:   %t/test.swift \
// RUN:   -plugin-path %t/plugins -Xfrontend -dump-macro-expansions \
// RUN:   -module-name TestModule \
// RUN:   == \
// RUN:   -shell -- echo '### 2' \
// RUN:   == \
// RUN:   -req=syntactic-expandmacro \
// RUN:   -req-opts=12:3:%t/bitwidthNumberedStructs.json \
// RUN:   -req-opts=2:1:%t/wrapAllProperties.json \
// RUN:   -req-opts=5:3:%t/myPropertyWrapper.json \
// RUN:   -req-opts=1:1:%t/DelegatedConformance.json \
// RUN:   %t/test.swift \
// RUN:   -- \
// RUN:   %t/test.swift \
// RUN:   -plugin-path %t/plugins -Xfrontend -dump-macro-expansions \
// RUN:   -module-name TestModule \
// RUN:   | tee %t.response

// RUN: diff -u %s.expected %t.response
