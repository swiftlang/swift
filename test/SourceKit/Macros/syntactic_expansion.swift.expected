### 1
source.edit.kind.active:
  1:1-1:22 ""
source.edit.kind.active:
  13:1-13:1 "static func requirement() where Element : P {
  Element.requirement()
}"
source.edit.kind.active:
  13:2-13:2 "extension Generic: P where Element: P {
}"
source.edit.kind.active:
  5:3-5:21 ""
source.edit.kind.active:
  7:17-7:17 "{
    get {
        _value.wrappedValue
      }
    set {
        _value.wrappedValue = newValue
      }
}"
source.edit.kind.active:
  7:17-7:17 "var _value: MyWrapperThingy<Int>"
source.edit.kind.active:
  2:1-2:19 ""
source.edit.kind.active:
  7:3-7:3 "@Wrapper"
source.edit.kind.active:
  10:3-10:3 "@Wrapper"
source.edit.kind.active:
  12:3-12:35 "struct blah8 {
  func __syntactic_macro_a01f07de3e9ee8585adf794ccd6dec8c6methodfMu_() {
  }
  func __syntactic_macro_a01f07de3e9ee8585adf794ccd6dec8c6methodfMu0_() {
  }
}
struct blah16 {
  func __syntactic_macro_a01f07de3e9ee8585adf794ccd6dec8c6methodfMu1_() {
  }
  func __syntactic_macro_a01f07de3e9ee8585adf794ccd6dec8c6methodfMu2_() {
  }
}
struct blah32 {
  func __syntactic_macro_a01f07de3e9ee8585adf794ccd6dec8c6methodfMu3_() {
  }
  func __syntactic_macro_a01f07de3e9ee8585adf794ccd6dec8c6methodfMu4_() {
  }
}
struct blah64 {
  func __syntactic_macro_a01f07de3e9ee8585adf794ccd6dec8c6methodfMu5_() {
  }
  func __syntactic_macro_a01f07de3e9ee8585adf794ccd6dec8c6methodfMu6_() {
  }
}"
### 2
source.edit.kind.active:
  12:3-12:35 "struct blah8 {
  func __syntactic_macro_a01f07de3e9ee8585adf794ccd6dec8c6methodfMu_() {
  }
  func __syntactic_macro_a01f07de3e9ee8585adf794ccd6dec8c6methodfMu0_() {
  }
}
struct blah16 {
  func __syntactic_macro_a01f07de3e9ee8585adf794ccd6dec8c6methodfMu1_() {
  }
  func __syntactic_macro_a01f07de3e9ee8585adf794ccd6dec8c6methodfMu2_() {
  }
}
struct blah32 {
  func __syntactic_macro_a01f07de3e9ee8585adf794ccd6dec8c6methodfMu3_() {
  }
  func __syntactic_macro_a01f07de3e9ee8585adf794ccd6dec8c6methodfMu4_() {
  }
}
struct blah64 {
  func __syntactic_macro_a01f07de3e9ee8585adf794ccd6dec8c6methodfMu5_() {
  }
  func __syntactic_macro_a01f07de3e9ee8585adf794ccd6dec8c6methodfMu6_() {
  }
}"
source.edit.kind.active:
  2:1-2:19 ""
source.edit.kind.active:
  7:3-7:3 "@Wrapper"
source.edit.kind.active:
  10:3-10:3 "@Wrapper"
source.edit.kind.active:
  5:3-5:21 ""
source.edit.kind.active:
  7:17-7:17 "{
    get {
        _value.wrappedValue
      }
    set {
        _value.wrappedValue = newValue
      }
}"
source.edit.kind.active:
  7:17-7:17 "var _value: MyWrapperThingy<Int>"
source.edit.kind.active:
  1:1-1:22 ""
source.edit.kind.active:
  13:1-13:1 "static func requirement() where Element : P {
  Element.requirement()
}"
source.edit.kind.active:
  13:2-13:2 "extension Generic: P where Element: P {
}"
